sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java

```diff
@@ -544,6 +560,68 @@ private static boolean areAllMappingsEqual(MetaData metaData, String... indices)
         return true;
     }
 
+    @VisibleForTesting
+    static String getRerouteIndex(RerouteAnalyzedStatement statement, Row parameters) throws SQLException {
+        ShardedTable shardedTable = statement.tableInfo();
+        if (shardedTable instanceof DocTableInfo) {
+            DocTableInfo docTableInfo = (DocTableInfo) shardedTable;
+            String indexName = docTableInfo.ident().indexName();
+            PartitionName partitionName = AlterTableAnalyzer.createPartitionName(statement.partitionProperties(),
+                docTableInfo, parameters);
+            if (partitionName != null) {
+                indexName = partitionName.asIndexName();
+            } else {
+                if (docTableInfo.isPartitioned()) {
+                    throw new SQLException("table is partitioned however no partition clause has been specified");
+                }
+            }
+
+            return indexName;
+        }
+
+        // Table is a blob table
+        assert shardedTable.concreteIndices().length == 1 : "table has to contain only 1 index name";
+        return shardedTable.concreteIndices()[0];
+    }
+
+    private CompletableFuture<Long> executeRerouteRequest(AllocationCommand command, boolean retryFailed) {
+        final CompletableFuture<Long> resultFuture = new CompletableFuture<>();
+
+        ClusterRerouteRequest request = new ClusterRerouteRequest();
+        request.add(command);
+        transportClusterRerouteAction.execute(request, new ActionListener<ClusterRerouteResponse>() {
+            @Override
+            public void onResponse(ClusterRerouteResponse clusterRerouteResponse) {
+                if (clusterRerouteResponse.isAcknowledged()) {
+                    resultFuture.complete(1L);
+                } else {
+                    resultFuture.complete(-1L);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                resultFuture.completeExceptionally(e);
+            }
+        });
+        return resultFuture;
+    }
+
+    public CompletableFuture<Long> executeRerouteMoveShard(RerouteMoveShardAnalyzedStatement statement, Row parameters) {
+        String indexName;
+        try {
+            indexName = getRerouteIndex(statement, parameters);
+        } catch (SQLException e) {
+            return CompletableFutures.failedFuture(e);
+        }
+        int shardId = ExpressionToNumberVisitor.convert(statement.shardId(), parameters).intValue();
+        String fromNodeId = ExpressionToStringVisitor.convert(statement.fromNodeId(), parameters);
```

Comment:
Is it safe to have those calls outside a try/catch block? If this method raises the error is probably not catched because it returns a future?